### ISSUE TITLE
Goldeneye Fog HLE issues

### ISSUE THREAD

**Issue by fzurita** [gonetz/GLideN64#2547]
This was first reported here: https://github.com/mupen64plus-ae/mupen64plus-ae/issues/926

Mainly, fog is not working correctly in goldeneye in HLE, but it works in LLE.

HLE
![Screenshot_20210805-193755_M64Plus AE](https://user-images.githubusercontent.com/11581687/128434808-7c7f9dba-41fe-496c-a539-6afff38f04b4.jpg)

LLE
![Screenshot_20210805-193723_M64Plus AE](https://user-images.githubusercontent.com/11581687/128434826-4fcc7cbb-32b6-4822-a6f2-9d1897e81147.jpg)

**Comment by gonetz** [comment]
It seems to be a problem with mobile version.
It looks like this on my PC with cdx4 hle:
![goldeneye-003](https://user-images.githubusercontent.com/8983092/131216726-f79543d4-be67-4d40-b88f-386efd01d7d0.jpg)

**Comment by fzurita** [comment]
Interesting, I went through all the settings and I found nothing that made the issue go away.

Blending seems to not be working correctly for the mountain texture. I know that for mobile we are doing something special here.

@standard-two-simplex any ideas? Do you think the blending changes I did here with your help are breaking this?

https://github.com/gonetz/GLideN64/pull/2372

**Comment by standard-two-simplex** [comment]
It's difficult to tell just by looking at the code.

Can you try forcing ordinary or legacy blending in `GraphicsDrawer::setBlendMode()`? If those are working, then it is a problem of the framebuffer fetch implementation.

Also, is it a regression or has it always been broken in android?

**Comment by fzurita** [comment]
Yeah, you are probably right. It really sounds like a ucode implementation issue.

**Comment by fzurita** [comment]
Glide64mk2 works correct, I'll compare the ucode implementation and see if there is anything significant there.

**Comment by weinerschnitzel** [comment]
I'm also curious about this being a regression. Glide64 differences may also be difficult leads. Does pr4 show proper result on mobile? If that much is narrowed, it would be easier to bisect and pinpoint when the regression happened for mobile and have a solid lead.

**Comment by fzurita** [comment]
Going back to ordinary blending doesn't fix it.

**Comment by fzurita** [comment]
@gonetz Any clues? How is fog different between HLE and LLE in this instance?

**Comment by Nejsplitter** [comment]
just to report, this issue also exists on the xbox series x version of retroarch/mupen/gliden64/hle, the mupen version is the gles3 and the issue is fixed when using cxd4 lle on there too.

**Comment by standard-two-simplex** [comment]
> It seems like fog is done outside the shader for LLE:
https://github.com/gonetz/GLideN64/blob/master/src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramUniformFactory.cpp#L385-L388

Yes, in LLE it seems that the fog parameter is precalculated by microcode and passed as alpha value of `aColor`.  I'm guessing this because HLE seems to overwrite it.

https://github.com/gonetz/GLideN64/blob/5aa389a86bb4b745eb2327584d382a0004ee3f0b/src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp#L320

In HLE, `aColor` doesn't have the shade alpha and it is calculated here. 

https://github.com/gonetz/GLideN64/blob/5aa389a86bb4b745eb2327584d382a0004ee3f0b/src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp#L340-L351

Please, try to hack different values from 0.0 to 1.0 into `vShadeColor.a` to confirm that it affects the scene. It does on desktop, at least.

**Comment by fzurita** [comment]
Yep, it has a very distinct effect on mobile. Is this happening because of perspective correction?

**Comment by weinerschnitzel** [comment]
Maybe it is discarded before it gets here in the shader, but drawn elsewhere? Maybe involving  software clipping? The mention of perspective correction reminds me of the noperspective shadecolor changes done here: https://github.com/gonetz/GLideN64/issues/2321#issuecomment-738733837

There are different branches for gles, does changing any of that code to follow gles path show the same behavior as pc? That said, I don't see how software vertex clipping could help unless it is right now being done at a different stage on android.

I'm think I'm a bit off base, anyways, I hope something here is related enough to give a hint to research further. Good luck!

**Comment by standard-two-simplex** [comment]
The usual fog formula is (assuming it has already been divided by w)
```
fog_factor = (z - z_min)/(z_max - z_min) 
```
which maps `z_min` to 0 and `z_max` to 1.

By defining `scale=1/(z_max - z_min)` and `offset=-z_min/(z_max-z_min)` you can simplify the formula above to,
```
fog_factor = z * scale + offset
```
which is computationally cheaper.

The game uses `scale=100` and `offset=-99` (passed to the shader as `uFogScale`) which correspond to `z_min=0.99` and `z_max = 1`. It is strange, I was expecting to find `scale` and `offset` in the `[0,1]` range and `z_min` and `z_max` not to be normalized.

> Also, uFogScale.t is always zero as well.

It is `-99.0` on desktop, you may dump it [here](https://github.com/gonetz/GLideN64/blob/e2ae3a3e4ec92c0ba464413850261641dffbaaa1/src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramUniformFactory.cpp#L399).

> So that fog is basically set by aPosition.z shows that aPositions.z is zero for all the bugged fog areas. I don't see how this could be any different in desktop, see this picture:

On desktop. `aPosition.z` is around 4000 to 8000 for the mountain. So is `aPosition.w` so the division falls between 0.99 and 1. I get a blue mountain with the change you proposed.

I doubt there is a difference between desktop and mobile outside of the shader. Are you sure you checked well? For example `if (fp < 0) fp = 1.0;` shoul be `if (fp < 0.0) fp = 1.0;`, so that the boolean evaluates well.

It looks like a precision issue. I notice that `fp` is declared low precision. Maybe there is an information loss when writing to it. Try declaring it medium or high precision.

**Comment by fzurita** [comment]
Increasing precision of ```fp``` to highp doesn't change the results. Interesting that you get different results there. So to me, it seems like the problem is outside the shader, like you said, since ```aPosition.z``` is not 0.0 for you at the mountain.

**Comment by standard-two-simplex** [comment]
> Also, uFogScale.t is always zero as well.

Uploading this value is straightforward. It is passed from `gSP.fog.offsetf` to `uFogScale.t`. Either there is a precision issue or you must have measured it wrongly. 

> So to me, it seems like the problem is outside the shader, like you said

Actually, I said the opposite, I believe that there is no problem outside the shader. In my opinion, the problem is most likely in the line
```
fp = aPosition.z/aPosition.w*uFogScale.s + uFogScale.t;	\n" 
```

**Comment by fzurita** [comment]
Sorry, I read too fast. Yeah, either I'm measuring wrong or the problem is outside the shader... I will find out soon.

**Comment by Jj0YzL5nvJ** [comment]
It's probably unrelated but I mention it just in case. Currently the OpenGL 3.3 support on NVIDIA card is broken. #2507

**Comment by standard-two-simplex** [comment]
@fzurita Any luck with this?

I came up with some tests to see if there is a precision issue. https://github.com/standard-two-simplex/GLideN64/tree/fog_test

Try the four commits in them. They should all give this picture. If some commits show this picture but others don't we will know the issue is here.

![goldeneye-001](https://user-images.githubusercontent.com/12202580/132946368-d9881c7d-0cd8-4674-adf6-9f2bea41cae1.png)

**Comment by fzurita** [comment]
I don't think it's a precision issue. I added this code:

```
			"      if (aPosition.z == 0.0) fp = 1.0;						\n"
```

I end up with this picture:
![goldeneye-000 (1)](https://user-images.githubusercontent.com/11581687/132954539-c2efaaed-88d6-4d3c-8baa-e6f108172df8.png)

So, the bug makes sense because:

0.0/aPosition.w*100 + (-99) == -99;
clamp(-99, 0.0, 1.0) == 0.0

**Comment by fzurita** [comment]
I put some logs int he buffered drawer: The vertex attributes do indeed have a lot of 0's for the z coordinate, these are typical values that I see:
```
2021-09-11 13:18:12.802 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(16.366983,-37.958527,0.000000,37.037018)"
2021-09-11 13:18:12.802 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(11.118143,-38.275360,0.000000,39.341812)"
2021-09-11 13:18:12.802 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(15.731347,-32.970455,0.000000,40.535110)"
2021-09-11 13:18:12.802 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(16.310602,-31.516047,0.000000,34.256790)"
2021-09-11 13:18:12.802 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(15.404597,-32.387325,0.000000,35.520767)"
2021-09-11 13:18:12.802 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(12.643794,-31.642906,0.000000,33.718918)"
2021-09-11 13:18:12.804 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(11.007145,-26.642569,0.000000,51.610062)"
2021-09-11 13:18:12.804 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(14.132325,-24.779678,0.000000,37.821854)"
2021-09-11 13:18:12.804 30852-31513/org.mupen64plusae.v3.alpha.debug E/GLideN64: opengl_UnbufferedDrawer.cpp:56,ERROR, "zero, coord=(12.349171,-24.228214,0.000000,51.268982)"
```

**Comment by fzurita** [comment]
Well, the problem lies here. Commenting out this code fixes the issue:
https://github.com/gonetz/GLideN64/blob/master/src/GraphicsDrawer.cpp#L99-L106

**Comment by fzurita** [comment]
@gonetz 

What is this for?
```
	case graphics::SpecialFeatures::ClipControl:
		return !m_glInfo.isGLESX;

```

**Comment by standard-two-simplex** [comment]
> Well, the problem lies here. Commenting out this code fixes the issue:

Nice finding.

The `git blame` of the blobs you posted points to old near clipping hacks.

https://github.com/gonetz/GLideN64/blame/e2ae3a3e4ec92c0ba464413850261641dffbaaa1/src/GraphicsDrawer.cpp#L99-L106

https://github.com/gonetz/GLideN64/blame/e2ae3a3e4ec92c0ba464413850261641dffbaaa1/src/Graphics/OpenGLContext/opengl_ContextImpl.cpp#L525-L526

Some of those commits are related to #1784.

```
if (GBI.isNoN() && gDP.otherMode.depthCompare == 0 && gDP.otherMode.depthUpdate == 0) { 
 		for (u32 i = firstIndex; i < triangles.num; ++i) { 
 			SPVertex & vtx = triangles.vertices[triangles.elements[i]]; 
 			vtx.z = 0.0f; 
 		} 
 	} 
}
```

This may no longer be necessary after software clipping and new near clipping hack. https://github.com/gonetz/GLideN64/issues/2427

@loganmc10  and @gonetz will know better.

**Comment by gonetz** [comment]
Interesting. I'll check it and try to remember.

**Comment by gonetz** [comment]
Ok. ClipControl means that OpenGL supports disabling of near-plane clipping (CLIP_DISTANCE0)
Only full GL supports it, so 
```
	case graphics::SpecialFeatures::ClipControl:
		return !m_glInfo.isGLESX;
```
This hack was introduced by me in commit d8817013cd1d0:
```
Fix geometry clipping problem for GL ES.
    
    NoN - no near clipping.
    GL ES does not support z-clipping disabling, so harmless hack used here.
    Harmlessness condition: z value is not used by pixel pipeline.
    gDP.otherMode.depthCompare == 0 && gDP.otherMode.depthUpdate == 0 - z not used for depth compare/update.
    Thus we can set vertex z to any value, which will ensure that z/w is in range [-1, 1], e.g. to zero.
    
    Fixed issue #588
```

So, this hack is not that harmless as I thought. And it only partially fixed the problem with near-plane clipping.
The software vertex clipping should completely eliminate the problem for both mobile and desktop GL.

Regarding WeakBlitFramebuffer. It is unrelated to this issue.
As I remember, GLES can't do blit frame buffer if source and destination buffers have different sizes.
Plugin works a bit less optimal with WeakBlitFramebuffer.

So, I propose to have software vertex clipping to be always enabled with GLES, that is hide this option option from user.
I planned to remove it at all, but there still can be issues with it, so this option helps to quickly find if some glitch is caused by software clipping. The hack, which is controlled by 
`if (!Context::ClipControl) { `
should be removed along with ClipControl flag and related code.
@fzurita, could you do it, test and make a PR?

**Comment by fzurita** [comment]
Yeah, sure.

**Comment by QmwJlHuSg9pa** [comment]
Is https://github.com/gonetz/GLideN64/issues/2333 possibly related?

Also, I'm having another issue with Jet Force Gemini that may be related? During cutscenes, the front edges of the camera avatar frustrum/matte box are visible on-screen (it appears as a grey box frame). Specifically, the front of this camera object: https://tcrf.net/File:JFG-Debugobjects3.png

Both issues are aarch64.

**Comment by gonetz** [comment]
> Is #2333 possibly related?

Hardly. #2333 is frame-buffer emulation related.



> Also, I'm having another issue with Jet Force Gemini that may be related?

May be. fzurita removed that near-clipping hack. The fix is in master. Please check, is that issue fixed as well?
If not, please open a new bug report, with a savestate if possible.

**Comment by fzurita** [comment]
Fixed by f2a69be96fd0394b4492e72204313b3f12ed00b8

**Comment by Nejsplitter** [comment]
@m4xw  any chance you could soon integrate this latest Gliden64 into the mupen angle build for uwp ?
